### PR TITLE
GH-39798: [C++] Optimize Take for fixed-size types including nested fixed-size lists

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

Introduce utilities for dealing with fixed-width types (including fixed-size lists of fixed-width types) generically. And use it for initial optimizations of `Take` and `Filter`.

### What changes are included in this PR?

- [x] Introduce utilities for dealing with fixed-width types generically
- [x] Use faster `Take` kernel on small power-of-2 byte widths of fixed-width types
  - [x] from `FSLTakeExec` (including FSLs of FSBs)
  - [x] from `FSBTakeExec` (done before this PR)
- [x] ~Take on any fixed-width type~ (as a separate issue #41301)
- [x] Use faster `Filter` kernel on both primitive and fixed-width types of any length
   - [x] from `FSLFilterExec` (including FSLs of FSBs)
   - [x] from `FSBFilterExec` (done before this PR)

### Are these changes tested?

By existing and new tests.

### Are there any user-facing changes?

Some functions added to the `arrow::util` namespace and documented inline.
* GitHub Issue: #39798

### CODE REVIEW

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (benibus) [inline cpp/src/arrow/util/fixed_width_internal.h:308]: What's the intention behind the `*Fallback` functions? Are there plans to introduce faster paths later or are they just meant to hide implementation details?

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (benibus) [inline cpp/src/arrow/util/fixed_width_internal.h:360]: Why use `assert` here instead of `DCHECK`?

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:308]: It's to allow the hot code path to be inlined while the cold code stays behind the function call.

```cpp
inline const uint8_t* OffsetPointerOfFixedWidthValues(const ArraySpan& source) {
  auto type_id = source.type->id();
  if (ARROW_PREDICT_TRUE(is_fixed_width(type_id))) {
    if (ARROW_PREDICT_FALSE(type_id == Type::BOOL)) {
      // BOOL arrays are bit-packed, thus a byte-aligned pointer cannot be produced in the
      // general case. Returning something for BOOL arrays that happen to byte-align
      // because offset=0 would create too much confusion.
      return nullptr;
    }
    return source.GetValues<uint8_t>(1, 0) + source.offset * source.type->byte_width();
  }
  return internal::OffsetPointerOfFixedWidthValuesFallback(source);
}
```

This is good for binary size while still allowing code like

```cpp
if (is_fixed_width(array.type->id())) {
  auto ptr = OffsetPointerOfFixedWidthValues(array);
  ptr->...
}
```

to lower to the equivalent of

```cpp
return source.GetValues<uint8_t>(1, 0) + source.offset * source.type->byte_width();
```

The binary size savings add up because the number of templates specializations using these functions is a proportional to `O(number of small powers-of-2 * index types * ...)` and replacing a `GetValues` call with `OffsetPointerOfFixedWidthValues` costs nothing for the common fixed-width types.

Comment (benibus) [reply cpp/src/arrow/util/fixed_width_internal.h:308]: Sounds good! Thanks for the explanation.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:360]: To not include `logging.h` in headers.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:308]: For instance, loop I just wrote:

```cpp
  for (size_t i = 0; i < num_chunks; ++i) {
    const ArraySpan chunk{*chunks[i]->data()};
    if (chunk.type->id() == Type::BOOL) {
      src_chunks_[i] = chunk.GetValues<uint8_t>(1, 0);
      src_offsets_[i] = chunk.offset;
    } else {
      src_chunks_[i] = util::OffsetPointerOfFixedWidthValues(chunk);
      src_offsets_[i] = 0;
    }
  }
```

File: cpp/src/arrow/util/fixed_width_test_util.h
Comment (bkietz) [inline cpp/src/arrow/util/fixed_width_test_util.h:60]: IIUC this could be simplified with
```c++
template <typename T>
Result<std::shared_ptr<Buffer>> IotaBuffer(int64_t length, T increment) {
  ARROW_ASSIGN_OR_RAISE(auto buf, AllocateBuffer(sizeof(T) * length));
  T value = 0;
  for (T& element : buf->mutable_span_as<T>()) {
    element = value;
    value += incrememnt; // handle overflow...
  }
  return buf;
}
```

This could be used to build the values in the non-list array, and to create the offsets for the nested list array.

File: cpp/src/arrow/util/fixed_width_test.cc
Comment (bkietz) [inline cpp/src/arrow/util/fixed_width_test.cc:117]: Should this also exercise dictionary?

Comment (bkietz) [reply cpp/src/arrow/util/fixed_width_test.cc:117]: ... and fixed size binary?

**Review by bkietz — CHANGES_REQUESTED** [review]
Thank you for working on this! My greatest concern is the number of header utilities introduced for handling fixed-width-like. It seems unnecessary to ensure that so many of these are inlined when they don't appear inside a hot loop.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:308]: I'm not inlining the entire function, just the cases where it lowers to a simple `GetValues` and exposing the conditionals so the compiler can avoid performing redundant checks where possible.

I understand it looks obsessive, but do you want me to put effort to undo this work?

Keeping the top-level checks in the header also serves as documentation.

```
const uint8_t* OffsetPointerOfFixedWidthValues(const ArraySpan& source) {
  auto type_id = source.type->id();
  if (ARROW_PREDICT_TRUE(is_fixed_width(type_id))) {
    if (ARROW_PREDICT_FALSE(type_id == Type::BOOL)) {
      // BOOL arrays are bit-packed, thus a byte-aligned pointer cannot be produced in the
      // general case. Returning something for BOOL arrays that happen to byte-align
      // because offset=0 would create too much confusion.
      return nullptr;
    }
    return source.GetValues<uint8_t>(1, 0) + source.offset * source.type->byte_width();
  }
  return internal::OffsetPointerOfFixedWidthValuesFallback(source);
}
```

This is how I use it in a loop

```cpp
  src_chunks_.resize(chunks.size());
  for (size_t i = 0; i < chunks.size(); ++i) {
    const ArraySpan chunk{*chunks[i]->data()};
    DCHECK(util::IsFixedWidthLike(chunk));
    if (chunk.type->id() == Type::BOOL) {
      src_chunks_[i] = chunk.GetValues<uint8_t>(1, 0);
    } else {
      src_chunks_[i] = util::OffsetPointerOfFixedWidthValues(chunk);
    }
  }
```

File: cpp/src/arrow/util/fixed_width_test.cc
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_test.cc:117]: I will add. But note that the algorithm only branches on `BOOL`, `is_fixed_width(t)`, and `FIXED_SIZE_LIST`.

File: cpp/src/arrow/util/fixed_width_test_util.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_test_util.h:60]: I don't understand. My functions calls `Append` on a values builder, yours allocates a `Buffer`. If I go this route, I will have to use very low-level code to build the other buffers of the lists. I would do that in production code, but this is test code.

**Comment by felipecrv** [comment]
This push was just a rebase. Fixes coming next.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:202]: I'm adding a version that takes `bool exclude_dictionary=false` and keeping the templated version because it's very possible that a kernel under development (unlike `Take`) doesn't handle all fixed-width types right away (e.g. it could be handling only the power-of-2 sizes and not the arbitrary-length types).

File: cpp/src/arrow/util/fixed_width_test.cc
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_test.cc:131]: Done.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:202]: Pushed.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h]: I pushed a commit removing all the inline code from the `.h` except for the template function.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (bkietz) [inline cpp/src/arrow/util/fixed_width_internal.h:249]: I expect most of the Windows failures https://github.com/apache/arrow/actions/runs/8929734699/job/24528249637?pr=41297 are due to the lack of _declspec(dllexport)
```suggestion
ARROW_EXPORT int64_t FixedWidthInBytes(const DataType& type);
```

File: cpp/src/arrow/util/fixed_width_test.cc
Comment (bkietz) [reply cpp/src/arrow/util/fixed_width_test.cc:131]: You're right, my concern isn't pertinent to this test: I was thinking that the fast path with PrimitiveFilterExec will also be applied to FSB which happens to have small power-of-2 size? For example, the UUID extension type?

**Review by bkietz — APPROVED** [review]
LGTM after the link errors are resolved

File: cpp/src/arrow/util/fixed_width_test.cc
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_test.cc:131]: A-ha. That's up to the kernel implementation. It could choose to always `memcpy` to a runtime-known `byte_width` or have specializations (still using `memcpy` that specializes) of the well-known power-of-2 lengths. That's what what `Take` will be doing: a switch on byte width with `case` 1/2/4...256 and a `default` that uses the `byte_width` variable.

**Comment by felipecrv** [comment]
@bkietz linker errors fixed.

**Comment by felipecrv** [comment]
CI failures are unrelated. I'm going to merge now, but more feedback can be posted here or on the follow-up PRs that use and extend this work.

**Comment by felipecrv** [comment]
@ursabot please benchmark lang=R

**Comment by felipecrv** [comment]
@ursabot please benchmark lang=C++

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (pitrou) [inline cpp/src/arrow/util/fixed_width_internal.h:273]: You need to add `\param` entries for all parameters if you have any of them. I'm expecting this docstring to fail on some compilers such as clang when the corresponding warnings are enabled.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (pitrou) [inline cpp/src/arrow/util/fixed_width_internal.h:220]: What does byte alignment have to do with it? This seems extremely specialized.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (pitrou) [inline cpp/src/arrow/util/fixed_width_internal.h:294]: Most if not all API functions operating on arrays take the offset into account, so adding "Offset" to the function name does not bring any additional information. `FixedWidthValuesPointer` would be shorter and more consistent with `MutableFixedWidthValuesPointer` below.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (pitrou) [inline cpp/src/arrow/util/fixed_width_internal.h:300]: Why is that? It would cost almost nothing to compute the proper physical offset.

File: cpp/src/arrow/util/fixed_width_test_util.h
Comment (pitrou) [inline cpp/src/arrow/util/fixed_width_test_util.h:1]: Can this file please be moved into `arrow/testing`, since those are testing utilities?

File: cpp/src/arrow/util/fixed_width_test_util.h
Comment (pitrou) [inline cpp/src/arrow/util/fixed_width_test_util.h:37]: Most function definitions here should be moved into a dedicated `.cc` file. This would also avoid including too much in this header file, such as `builder.h` and `builder_primitive.h`.

**Review by pitrou — COMMENTED** [review]
Sorry for the belated review here.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:186]: I didn't feel comfortable having `DCHECK(IsFixedWidthLike(array));` changing `null_count` from `kUnknownNullCount` to `0` or the actual null count depending on whether it's a debug build or not. And since this predicate is used for routing kernel calls, I wanted to leave it up to the implementer to decide if it's worth paying the bitmap scan cost of `GetNullCount()` at every inner array.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:207]: I started without `BOOL` then almost every `IsFixedWidthLike` check in the `Take` kernel had to have `|| type->id() == BOOL`.

Additionally, excluding `BOOL` would put this function in conflict with `is_fixed_width()`. I wanted to preserve the property that the set of arrays that passes this predicate is a superset of the arrays that can pass `is_fixed_width(type->id())`. How could an array be `[Is]FixedWidthLike` if it's not even `[is_]fixed_width()`?

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:220]: I explained this in the long comment above. If I were to include BOOL in this, I would have to measure everything in bits and would not be able to produce a base pointer that can be written to with `memcpy` as I do in `OffsetPointerOfFixedWidthValues`.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:275]: That's what I'm solving soon, but I couldn't do it easily because the allocators in `KernelContext` zero-initialize the buffers unlike the allocators in the `MemoryPool`.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:294]: The inconsistency with `MutableFixedWidthValuesPointer` was on purpose because that one requires the input to have `offset==0`.

> Most if not all API functions operating on arrays take the offset into account...

And the pointers they return can easily have the offset applied to them after the fact. I gave the `Offset-` prefix to this one so it works as a reminder to the caller that `ptr + arr.offset` won't work (it might by accident), but it's not correct for the general case. The actual offset is a complicated sum of offset products.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:300]: Yes, it would be the same code I wrote for doing just that in `OffsetPointerOfFixedWidthValues` but it seemed overkill to bother with offsets on output arrays allocated by `PreallocateFixedWidthArrayData`. This `offset==0` requirement also makes it possible to use this on `BOOL` arrays.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:37]: Isn't that what `\brief` is for? It's rather long because of the examples.

Without them, it's really hard to understand the intricacies of this concept.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:300]: If I extend these functions to support bit-sized types I will write a function that returns an (offset, ptr) pair instead of just a pointer.

File: cpp/src/arrow/util/fixed_width_test_util.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_test_util.h:1]: Opened #41558

File: cpp/src/arrow/util/fixed_width_test_util.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_test_util.h:37]: Will do that in #41558 as well.

**Comment by felipecrv** [comment]
@pitrou replied to your comments, please resolve the ones that you think are acceptable and discuss further the ones you still think need to be addressed.

Many questions revolve around the handling of `BOOL` and offsets. Do you think I should extend this to support `BOOL` as well? With all the extra complexity that would bring?

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (pitrou) [reply cpp/src/arrow/util/fixed_width_internal.h:207]: My "weird API design choice" comment was about the "if not nested under FSL" part :-)

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (pitrou) [reply cpp/src/arrow/util/fixed_width_internal.h:37]: `\brief` is for a one-liner introductory sentence, but if every non-trivial API function gets a 150-line docstring, then nobody will read them anymore (and, besides, even the source code will become impossible to navigate).

Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:37]: I will remove the comments from the `///` and spread the explanation, but I think this concept is complicated enough that having a long comment in the source code is worth it.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:275]: I will fix this in one of the `Take` PRs as I want to ensure we zero the values only once.

File: cpp/src/arrow/util/fixed_width_internal.h
Comment (felipecrv) [reply cpp/src/arrow/util/fixed_width_internal.h:207]: API change in #41597.

**Pull request merged**
